rpcs3/Emu/Cell/SPUAnalyser.h

```diff
@@ -271,30 +275,48 @@ struct spu_function_t
 	bool does_reset_stack;
 
 	// Pointer to the compiled function
-	u32(*compiled)(SPUThread* _spu, be_t<u32>* _ls) = nullptr;
+	CompiledFunc compiled = nullptr;
 
-	spu_function_t(u32 addr, u32 size)
-		: addr(addr)
-		, size(size)
+	spu_function_contents_t(u32 addr, u32 size)
+		: addr(addr),
+		  size(size)
 	{
 	}
 };
 
+// A single instance of a compiled function, currently in use
+union spu_function_t
+{
+	// The function itself and its data
+	std::shared_ptr<spu_function_contents_t> contents;
+
+	// Whether pages the function is in were written to since its last execution
+	bool dirty_bit : 1;
+
+	operator bool()
+	{
+		return contents != nullptr;
+	}
+
+	spu_function_t() : contents(nullptr) {};
+	~spu_function_t() {dirty_bit = false; contents.reset();};
```

Comment:
Pretty sure that this thing is one big UB.

Reply:
True, it isn't called, but reset() is used instead of the destructor though.
Wanted it to be a union to it will be 8 bytes instead of 16. Considering the very frequent iteration of it, it will be much more friendly to the cache.
Making this not UB is easily possible with our own implementation of a shared_ptr, though.

Reply:
shared_ptr is 16 bytes in size

Reply:
@Farseer2 well, also because layout of shared_ptr is not defined and you don't know whether the bit you "unite" with is "free". This is very fragile. Ideally some tagged pointer class (shouldn't be hard to write) should be used here if 8 byte savings is the goal.